### PR TITLE
Replace cachegen kernels with more performant coder kernels

### PR DESCRIPTION
**Pull request by colinreyn**
https://github.com/LMCache/LMCache-Ascend/pull/143 wired up a series of cachegen kernels. They were however limited by the interface imposed by LMCache that invoked an expensive `RepeatInterleave` in encode. This PR replaces those kernels with more performant kernels that remove the need for the work, done in `collect_bytes`, that was previously imposing a bottle neck. An incisive change is injected into path of LMCache's cachegen encode/decode to properly invoke the new kernels

These kernels are in-themselves faster (~4x) and on the encode side remove other expensive operations leading to a substantial performance improvement.

Here are some representative timings showing a few key results:
- Cachegen decode is comparable to a naive serde with a remote backend (model and bandwidth dependenant)
- Cachegen encode is slower by 30 - 50% compared to naive
- The compression ratio ranges from 3.5x - 6x depending strongly on chunk size

image

image

image

image

A minimal drop in accuracy is measured against the gsm8k benchmark although it should be noted that these results have been observed to depend on chunk size and model

```
Qwen3-8B - No Cache
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8817|±  |0.0089|
|     |       |strict-match    |     5|exact_match|↑  |0.8772|±  |0.0090|

Qwen3-8B - Pure cachegen cache
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8560|±  |0.0097|
|     |       |strict-match    |     5|exact_match|↑  |0.8605|±  |0.0095|
```

```
Qwen3-30B-A3B - No Cache
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8423|±  |0.0100|
|     |       |strict-match    |     5|exact_match|↑  |0.8825|±  |0.0089|

Qwen3-30B-A3B - Pure cachegen cache
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8400|±  |0.0101|
|     |       |strict-match    |     5|exact_match|↑  |0.8431|±  |0.0100|
```
```
Qwen2.5-7B-Instruct - No Cache
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8196|±  |0.0106|
|     |       |strict-match    |     5|exact_match|↑  |0.7877|±  |0.0113|

Qwen2.5-7B-Instruct - Cacgen Cache
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6710|±  |0.0129|
|     |       |strict-match    |     5|exact_match|↑  |0.6399|±  |0.0132|
```

### CODE REVIEW

File: lmcache_ascend/serde/pac.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        lmc_ops.pac_encode(tmp_in, meta_data, local_out_buf, local_output_lengths)
+        max_len = local_output_lengths[-1, -1]
+
+        data_chunks.append(
+            CacheGenGPUBytestream(
+                bytestream=local_out_buf.flatten()[0:max_len],
+                bytestream_lengths=local_output_lengths,
+                ntokens=end - start,
+            )
+        )
```
[the remark this answers is not in the document]
Comment (colinreyn) [reply lmcache_ascend/serde/pac.py:133]: Gemini is wrong here about cloning the slice for tmp_in . `input_sym.sizes()` will return different results depending on if the `clone` is present or not. The number of tokens is detected from this so removing the clone would require broader changes. As the copy is not actually a performance bottleneck I'm ignoring that part of this suggestion. (moving the other clones is done)

File: csrc/pac_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+#include <torch_npu/csrc/core/npu/NPUStream.h>
+#include <torch_npu/csrc/framework/OpCommand.h>
+
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
+
+static constexpr uint32_t AIV_MAX = 20;
+
+void pac_perpare_enc_metadata(const at::Tensor &input_sym,
```
[the remark this answers is not in the document]
Comment (colinreyn) [reply csrc/pac_kernels.cpp:17]: Ack

File: lmcache_ascend/serde/pac.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    new_key, max_tensors_key = torch_quant_vectorized(key_bins, fp_k)
+    new_value, max_tensors_value = torch_quant_vectorized(value_bins, fp_v)
+    encode_input = torch.cat((new_key, new_value), dim=0).reshape(
+        nlayers, n_tokens, nchannels
+    )
+
+    meta_data = torch.zeros((nlayers, nchannels, n_bins), dtype=torch.int16).to(
+        device="npu"
+    )
```
[the remark this answers is not in the document]
Comment (colinreyn) [reply lmcache_ascend/serde/pac.py:104]: Ack

File: lmcache_ascend/serde/pac.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    lmc_ops.pac_perpare_enc_metadata(encode_input, meta_data)
+
+    output_buffer = torch.zeros((nlayers, nchannels, chunk_size), dtype=torch.uint8).to(
+        device="npu"
+    )
+    output_lengths = torch.zeros((nlayers, nchannels), dtype=torch.int32).to(
+        device="npu"
+    )
```
[the remark this answers is not in the document]
Comment (colinreyn) [reply lmcache_ascend/serde/pac.py:113]: Ack

File: csrc/pac_kernels.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  const c10::OptionalDeviceGuard device_guard(device_of(input_sym));
+  const aclrtStream stream = c10_npu::getCurrentNPUStream().stream();
+
+  auto meta_data_data_ptr = static_cast<uint8_t *>(meta_data.data_ptr());
+  auto input_data_ptr = static_cast<uint8_t *>(input_sym.data_ptr());
+  auto output_data_ptr = static_cast<uint8_t *>(output_buffer.data_ptr());
+  auto output_lengths_data_ptr =
+      static_cast<uint8_t *>(output_lengths.data_ptr());
+
+  auto workGM = torch::zeros({40 * 32}, input_sym.options().dtype(torch::kI32));
```
[the remark this answers is not in the document]
Comment (colinreyn) [reply csrc/pac_kernels.cpp:100]: Perhaps, but it looks to be around 0.5% (30us vs 6ms) of the overhead and, implemented as it is, removes any concerns around concurrency or generality. I'm inclined to keep this simple and pay the small overhead cost

File: lmcache_ascend/serde/pac.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+        output = output.permute(1, 0, 2, 3)
+        output = output.reshape((2, layers_in_key, n_chunks * chunk_size, nchannels))
+        output = output[:, :, 0:n_tokens, :]  # Trim output down to n_tokens
+
+        out = output.reshape((2, layers_in_key, n_tokens, nchannels))
+        key, value = out.float()
+
+    e = decode_stream.record_event(e)
+    e.synchronize()
```
[the remark this answers is not in the document]
Comment (colinreyn) [reply lmcache_ascend/serde/pac.py:70]: The AI proposed fix isn't workable but revisiting the problem the sync is fixing, actually it might be addressable by adding some (seemingly) redundant copies instead. First indication is that it works and is more efficient

[the remark this answers is not in the document]
Comment (colinreyn) [reply lmcache_ascend/serde/pac.py:70]: First indication was wrong, correctness issues come back in - in principle I agree that this isn't perfect but practically speaking it works

File: csrc/pac_kernels.cpp [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  const int nchannels = input_shape[2];
+  const int nbins = meta_data.sizes()[2];
+
+  TORCH_CHECK(meta_data.device().is_privateuseone());
+  TORCH_CHECK(input_sym.device().is_privateuseone());
+  TORCH_CHECK(nchannels <= 4096,
+              "Number of channels exceeds that supported be encode, contact "
+              "LMCache Ascend about changing this limitation");
+  TORCH_CHECK(nbins <= 32,
+              "Number of bins exceeds that supported be encode, contact."
```
Comment (matthewygf) [inline csrc/pac_kernels.cpp]: nit: contact. -> contact

File: csrc/pac_kernels.cpp [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+              "LMCache Ascend about changing this limitation");
+  TORCH_CHECK(
+      nchannels % 32 == 0,
+      "Encode implementation relies on the number of channels being a multiple "
+      "of 32, contact LMCache Ascend about changing this limitation");
+
+  const c10::OptionalDeviceGuard device_guard(device_of(input_sym));
+  const aclrtStream stream = c10_npu::getCurrentNPUStream().stream();
```
Comment (matthewygf) [inline csrc/pac_kernels.cpp]: nit: replace `stream() -> stream(false)` if do not require synchronization of the task queue.

File: csrc/pac_kernels.cpp [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+  TORCH_CHECK(output_lengths.device().is_privateuseone());
+  TORCH_CHECK(nchannels <= 4096,
+              "Number of channels exceeds that supported be encode, contact "
+              "LMCache Ascend about changing this limitation");
+  TORCH_CHECK(ntokens <= 256,
+              "Number of tokens exceeds that supported be encode, handle by "
+              "chunking the input."
+              "Contact LMCache Ascend about changing this limitation");
+  TORCH_CHECK(nbins <= 32,
+              "Number of bins exceeds that supported be encode, contact."
```
Comment (matthewygf) [inline csrc/pac_kernels.cpp]: nit: contact. -> contact

File: csrc/pac_kernels.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (matthewygf) [reply csrc/pac_kernels.cpp]: As discussed, worth changing the Int8 instead of Int32.

File: lmcache_ascend/__init__.py [hunk 10]
```diff
@@ -262,6 +262,56 @@ def _patch_cacheblend():
...
+    # copy over the data
+    old_batched_get_blocking = RemoteBackend.batched_get_blocking
+
+    def new_batched_get_blocking(
+        self,
+        keys: List[CacheEngineKey],
+    ) -> List[Optional[MemoryObj]]:
+        source_bufs = old_batched_get_blocking(self, keys)
+
+        allocator = self.get_allocator_backend()
```
Comment (matthewygf) [inline lmcache_ascend/__init__.py]: As discussed, please only target cachegen backend for now.

**Review by matthewygf — CHANGES_REQUESTED** [review]

File: lmcache_ascend/serde/pac.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthewygf) [reply lmcache_ascend/serde/pac.py:61]: okay, leaving it as is.

File: csrc/pac_kernels.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (colinreyn) [reply csrc/pac_kernels.cpp]: Done

File: lmcache_ascend/__init__.py (hunk 10 above)
Comment (colinreyn) [reply lmcache_ascend/__init__.py]: Done

**Review by matthewygf — APPROVED** [review]
LGTM!
